src/core/ext/filters/client_channel/resolver/custom/custom_resolver.cc

```diff
@@ -0,0 +1,363 @@
+      auto next_completion = next_completion_;
+      next_completion_ = nullptr;
+      GRPC_CLOSURE_SCHED(next_completion, GRPC_ERROR_NONE);
+    }
+  }
+}
+
+void CustomResolver::ShutdownLocked() {
+  vtable_.destroy(user_data_);
+  if (next_completion_ != nullptr) {
+    *target_result_ = nullptr;
+    auto next_completion = next_completion_;
+    next_completion_ = nullptr;
+    GRPC_CLOSURE_SCHED(next_completion, GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+                                            "Resolver Shutdown"));
+  }
+}
+
+//
+// Observer
+//
+
+class CustomResolverObserver : public grpc_resolver_observer {
+ public:
+  CustomResolverObserver(const ResolverArgs& args,
+                         RefCountedPtr<Resolver> resolver) {
+    resolver_ = std::move(resolver);
+    channel_args_ = grpc_channel_args_copy(args.args);
+  }
+
+  ~CustomResolverObserver() { grpc_channel_args_destroy(channel_args_); }
+
+  void SetAddresses(grpc_lb_addresses* addresses) {
+    grpc_arg new_arg = grpc_lb_addresses_create_channel_arg(addresses);
+    auto channel_args =
+        grpc_channel_args_copy_and_add(channel_args_, &new_arg, 1);
+    auto resolver = static_cast<CustomResolver*>(resolver_.get());
+    resolver->SetNextResult(channel_args);
+  }
+
+ private:
+  RefCountedPtr<Resolver> resolver_;
+  grpc_channel_args* channel_args_ = nullptr;
+};
+
+//
+// Factory
+//
+
+class CustomResolverFactory : public ResolverFactory,
+                              public grpc_resolver_factory {
+ public:
+  CustomResolverFactory(const char* scheme, void* user_data,
+                        const ResolverFactoryVTable& factory_vtable,
+                        const ResolverVTable& resolver_vtable)
+      : scheme_(gpr_strdup(scheme)),
+        user_data_(user_data),
+        factory_vtable_(factory_vtable),
+        resolver_vtable_(resolver_vtable) {}
+
+  ~CustomResolverFactory() {
+    gpr_free(scheme_);
+    factory_vtable_.destroy(user_data_);
+  }
+
+  OrphanablePtr<Resolver> CreateResolver(
+      const ResolverArgs& args) const override {
+    auto resolver = New<CustomResolver>(args, resolver_vtable_);
+    auto observer = New<CustomResolverObserver>(args, resolver->Ref());
+    grpc_resolver_args api_args = {&args};
+    auto user_data = factory_vtable_.resolve(user_data_, &api_args, observer);
+    if (user_data == nullptr) {
+      resolver->Unref();
+      Delete(observer);
+      return nullptr;
+    }
+    resolver->Init(user_data);
+    return OrphanablePtr<Resolver>(resolver);
+  }
+
+  const char* scheme() const override { return scheme_; }
+
+ private:
+  char* scheme_;
+  void* user_data_;
+  ResolverFactoryVTable factory_vtable_;
+  ResolverVTable resolver_vtable_;
+};
+
+}  // namespace
+
+}  // namespace grpc_core
+
+grpc_uri* grpc_resolver_args_get_target(grpc_resolver_args* args) {
+  GPR_ASSERT(args != nullptr);
+  return args->args->uri;
+}
+
+grpc_addresses* grpc_addresses_create_with_capacity(size_t capacity,
```

Comment:
The `grpc_addresses_*` functions should live in lb_policy_factory.h, where the `grpc_lb_addresses` struct (and its methods) are defined.